### PR TITLE
feat: sns to lambda stage

### PR DESCRIPTION
**Pull request by malachi-constant**
### Relates
- #250 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
Integration test passing.

Screen Shot 2023-02-08 at 11 43 36 AM

**Comment by anmolsgandhi** [comment]
Continuing the conversation on redrive, i believe users can have multiple ways to redrive, most common i have seen is having a lambda to reprocess, that is also a pattern we have in SDLF. So essentially DLQ --> Lambda --> SNS event/Lambda logic --> SNS

**Review by anmolsgandhi — COMMENTED** [review]
Continuing the conversation on redrive, i believe users can have multiple ways to redrive, most common i have seen is having a lambda to reprocess, that is also a pattern we have in SDLF. So essentially DLQ --> Lambda --> SNS event/Lambda logic --> SNS

**Comment by anmolsgandhi** [comment]
> > Continuing the conversation on redrive, i believe users can have multiple ways to redrive, most common i have seen is having a lambda to reprocess, that is also a pattern we have in SDLF. So essentially DLQ --> Lambda --> SNS event/Lambda logic --> SNS
> 
> A Lambda could solve this, yes. The reason I think an SQS queue is still preferable is that it comes with [built-in redrive functionality](https://aws.amazon.com/blogs/compute/introducing-amazon-simple-queue-service-dead-letter-queue-redrive-to-source-queues/). As this is a relatively recent addition (December 2021), the Lambda pattern in SDLF was designed prior to this.
> 
> Another advantage to this is that a Lambda can only redrive 10 messages at a time per each call to SQS. This redrive feature in SQS allows you to redrive thousands of messages in a matter of seconds by just pressing the "redrive" button.

Yes that does make sense, if we move towards a SNS->SQS->LAMBDA stage then we completely dont provide for the SNS->LAMBDA usecase. thoughts ??

**Comment by malachi-constant** [comment]
> > > Continuing the conversation on redrive, i believe users can have multiple ways to redrive, most common i have seen is having a lambda to reprocess, that is also a pattern we have in SDLF. So essentially DLQ --> Lambda --> SNS event/Lambda logic --> SNS
> > 
> > 
> > A Lambda could solve this, yes. The reason I think an SQS queue is still preferable is that it comes with [built-in redrive functionality](https://aws.amazon.com/blogs/compute/introducing-amazon-simple-queue-service-dead-letter-queue-redrive-to-source-queues/). As this is a relatively recent addition (December 2021), the Lambda pattern in SDLF was designed prior to this.
> > Another advantage to this is that a Lambda can only redrive 10 messages at a time per each call to SQS. This redrive feature in SQS allows you to redrive thousands of messages in a matter of seconds by just pressing the "redrive" button.
> 
> Yes that does make sense, if we move towards a SNS->SQS->LAMBDA stage then we completely dont provide for the SNS->LAMBDA usecase. thoughts ??

My vote would be to add `SNS` as an optional feature to our existing sqs to lambda stage or just inherit from the existing stage with the additional functionality and event target.

**Comment by kukushking** [comment]
> if we move towards a SNS->SQS->LAMBDA stage then we completely dont provide for the SNS->LAMBDA usecase.

My 2 cents: using SNS + SQS together is a very common scalable pattern and it is able to cover same requirements as direct SNS -> lambda and SQS + lambda. You can scale consumers and producers independently and add more queues if running into lambda concurrency limits.

> The reason I think an SQS queue is still preferable is that it comes with [built-in redrive functionality](https://aws.amazon.com/blogs/compute/introducing-amazon-simple-queue-service-dead-letter-queue-redrive-to-source-queues/). As this is a relatively recent addition (December 2021), the Lambda pattern in SDLF was designed prior to this.

> Another advantage to this is that a Lambda can only redrive 10 messages at a time per each call to SQS. This redrive feature in SQS allows you to redrive thousands of messages in a matter of seconds by just pressing the "redrive" button.

Neat, didn't know about this addition. It was always a pain to only be able to redrive by 10 messages at a time.

**Review by anmolsgandhi — APPROVED** [review]

File: src/stages/sns-lambda.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  readonly targets?: events.IRuleTarget[];
+  readonly eventPattern?: events.EventPattern;
+
+  readonly topic: sns.ITopic;
+  readonly snsDeadLetterQueue?: sqs.Queue;
+
+  constructor(scope: Construct, id: string, props: SnsToLambdaStageProps) {
+    super(scope, id, props);
+
+    this.snsDeadLetterQueue = props.snsDlqEnabled ? new sqs.Queue(this, "SNS Dead Letter Queue", {}) : undefined;
```
[the remark this answers is not in the document]
Comment (malachi-constant) [reply src/stages/sns-lambda.ts:29]: The SNS subscription to SQS itself has an option to provide a DLQ queue. Not sure if that would be completely redundant. That would be messages specifically which fail posting to the queue for whatever reason.

**Review by LeonLuttenberger — APPROVED** [review]

**Pull request merged**